README.md
[![Travis Build](https://travis-ci.org/Ks89/My-MEAN-website-server.svg?branch=master)](https://travis-ci.org/Ks89/My-MEAN-website-server)   [![AppVeyor Build](https://ci.appveyor.com/api/projects/status/x7r2v139hi84cvsj/branch/master?svg=true)](https://ci.appveyor.com/project/Ks89/my-mean-website-server/branch/master)   [![Code Climate](https://codeclimate.com/github/Ks89/My-MEAN-website-server/badges/gpa.svg)](https://codeclimate.com/github/Ks89/My-MEAN-website-server)   [![CodeClimate Coverage](https://codeclimate.com/github/Ks89/My-MEAN-website-server/badges/coverage.svg)](https://codeclimate.com/github/Ks89/My-MEAN-website-server/coverage)   [![Coveralls Coverage](https://coveralls.io/repos/github/Ks89/My-MEAN-website-server/badge.svg?branch=master)](https://coveralls.io/github/Ks89/My-MEAN-website-server?branch=master)   [![Known Vulnerabilities](https://snyk.io/test/github/ks89/my-mean-website-server/badge.svg)](https://snyk.io/test/github/ks89/my-mean-website-server)


# My M.E.A.N. website server/back-end (Alpha)


**This is the server side.** Client side is available [HERE](https://github.com/Ks89/My-MEAN-website-client)


## Informations

My MEAN website is a MEAN's web application that I'm creating as a personal website, but also for other users.
It's composed by:

- M: a MongoDb's database
- E: a back-end with Express js
- A: a front-end in Angular 2
- N: a back-end in Node.js
- redis
- webpack
- and other stuff

A possible extension of this project will be a configurable template to build a custom web app very quickly. This is my final goal, please be patient :)

Attention! This project is still an alpha, so it's not production ready. Please be careful.
If you are interested, star this project on GitHub, share it and create pull requests.

Testing:

- back-end unit: almost done (only the necessary things)*. coverage >=90%
- back-end integration: almost done*. coverage >=90%

(*) I unit-tested only public functions and I tested all APIs (integration) except for OAUTH2/PassportJS.
This is because, it's extremely difficult to test passportjs (for 3dparty services, not for the local auth) without to use browsers (like Zombie or Phantom). In my opinion an integration-test for a back-end api must use only backend's code, not also a browser (browser is on client and not on server :) ).
The problem is that to test PassportJS without a browser it's really difficult. I asked on StackOverflow [HERE](http://stackoverflow.com/questions/38169351/how-can-i-test-integration-testing-with-supertest-a-node-js-server-with-passpo), without receive any answers.
For this reason, I decided to unit-tests these APIs (not APIs theirself but their functions/logics).
If you want to help me to write integration-test's case for PassportJS, check [this file](https://github.com/Ks89/My-MEAN-website-server/blob/master/test-server-integration/TODO-auth-3dparty.experimentalspec.js)

## Requirements

- macOS, Linux or Windows 10 **with admin privileges**
- Node.js + npm
- MongoDB and redis
- some global npm dependencies
- work in progress... (this is only an alpha, please be patient)

## News

- *04/09/2017* - **My MEAN website** Alpha 6 public release [HERE](https://github.com/Ks89/My-MEAN-website-server/releases/tag/v.alpha-6)
- *01/21/2017* - **My MEAN website** Alpha 5 public release [HERE](https://github.com/Ks89/My-MEAN-website-server/releases/tag/v.alpha-5)
- *12/30/2016* - **My MEAN website** Alpha 4 public release [HERE](https://github.com/Ks89/My-MEAN-website-server/releases/tag/v.alpha-4)
- *11/28/2016* - **My MEAN website** Alpha 3 public release [HERE](https://github.com/Ks89/My-MEAN-website-server/releases/tag/v.alpha-3)
- *10/27/2016* - **My MEAN website** Alpha 2 public release [HERE](https://github.com/Ks89/My-MEAN-website-server/releases/tag/v.alpha-2.2)
- *08/15/2016* - **My MEAN website** Alpha 1 public release [HERE](https://github.com/Ks89/My-MEAN-website-server/releases/tag/v.alpha-1)


## How to install (MacOS)

- from the `setup` folder of this project, run `bash install-macos.sh`
- import the db dump (.json) from `docs`'s folder using MongoChef or another software

## How to install (Linux)

- from the `setup` folder of this project, run `bash install-linux.sh`
- import the db dump (.json) from `docs`'s folder using MongoChef or another software

## How to install (Windows)

*Tested only on Windows 10*

- install both [Mozilla Firefox](https://www.mozilla.org/en-US/firefox/new/) and [Google Chrome](https://www.google.com/chrome/browser/desktop/index.html)
- install Node.js from the [official website](https://www.nodejs.org)
- install MongoDb Community from the [official website](https://www.mongodb.com)
- create a db called `KS` (obviously you have to start MongoDb to do that). You can configure this value later, but for debug/local environment I suggest this name
- **import the db dump (.json) from `docs`'s folder using MongoChef** or another software [HERE](http://3t.io/mongochef/download/) (obviously you have to start MongoDb to do that)
- install redis-server for Windows (file .msi) [HERE](https://github.com/MSOpenTech/redis/releases)
- install Python 2.7.x from the [official website](https://www.python.org)
- from the `setup` folder of this project, run with PowerShell as administator `bash install-windows.sh`

If you'll have problems with `node-zopfli`, you have to install it properly following [this tutorial](https://github.com/nodejs/node-gyp#installation). There are two options, try with the first one `npm install --global --production windows-build-tools`, if it will fail, use option 2.
Both options will require to download really big files from microsoft.com (manually or automatically). So, be careful.

## How to setup

1. You have to rename `.env_example`'s file into `.env` (debug/local config - mandatory) and create another copy called `.env_prod` (production config - to deploy)<br>
1a. Configure `FRONT_END_PATH` with the relative position of the main folder of [THE CLIENT SIDE OF THIS PROJECT](https://github.com/Ks89/My-MEAN-website-client). For `.env_prod` you should use simply `public`<br>
1b. Configure `REDIS_HOST`, `REDIS_PORT` and `REDIS_TTL`. If you are in a debug/local environment you should use the example values<br>
1c. Configure `MONGODB_URI` with your db path. I suggest to use the example value without user/password and with a db called `KS`<br>
1d. replace `YOU KEY/ID` with the keys obtained from facebook/github... oauth applications.<br>
1e. replace `YOUR_EMAIL` and `YOUR_PASSWORD` with the data of your e-mail account<br>
1f. replace `YOUR GOOGLE RECAPTCHA...` with Google Recaptcha2's keys<br>
1g. reaplace `INSERT A JWT SECRET HERE` with an alphanumerical string (I'm using a random string with a length = 72)
2. execute this command `npm install` into the root folder  (if it will fail, run it again :))
3. start redis-server (both on Linux and Mac run `redis-server`, on Windows start `C:\Program files\Redis\redis-server.exe`)
4. start MongoDb (on Linux run `mongod`, on Mac run `sudo mongod` and on Windows start `C:\Program Files\MongoDB\Server\3.4\bin\mongod.exe`)
5. execute this command `npm start` into the root folder to start this application (back-end)

## How to run tests (server-side)

- `npm test`

## How to start (development mode)

- cd 'main folder of this project'
- `npm start`
- all REST webservices will be available at http://localhost:3000 for instance try with http://localhost:3000/api/projects

## How to start (production mode)

*I decide to use forever to run this project in a production environment*

- cd 'main folder of this project'
- `npm run prod:start`
- all REST webservices will be available at http://localhost:3000 for instance try with http://localhost:3000/api/projects
- if you want to stop it use `npm run prod:stop`

## Features

Work in progress... (this is only an alpha, please be patient)

## Future extensions

Work in progress... (this is only an alpha, please be patient)

## Images

![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/home.png)


![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/projects.png)


![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/projectDetail.png)


![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/projectDetail-image.png)


![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/contact.png)


![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/contact-images.png)


![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/signin.png)


![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/register.png)


Note: updated local profile info
![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/profile-updated.png)

Note: multiple account connected (Facebook and Github)
![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/profile-multiple.png)

## Configuration

From Alpha 5 I created many entries in .env, so you could start to configure this application.
Work in progress... (this is only an alpha, please be patient)

## Thanks

A special thanks to the authors of this book, because very useful to understand how to develop a modern web application: [BOOK1](https://www.manning.com/books/getting-mean-with-mongo-express-angular-and-node)

## License

Copyright 2015-2017 Stefano Cappa

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

<br/>
**Created by Stefano Cappa**


NEWS.md
For news, upcoming features and so on, refers to [THIS PAGE](https://github.com/Ks89/My-MEAN-website-server/milestones)

app.js
'use strict';
const config = require('./src/config');
let logger = require('./src/utils/logger-winston.js');

logger.warn('Starting with NODE_ENV=' + config.NODE_ENV);
logger.warn('config.CI is ' + config.CI);

if (!config.isCI()) {
  console.log('Initializing dotenv (requires .env file)');
  if (config.isProd()) {
    // production
    require('dotenv').config({path: '.env_prod'}); //to read info from .env_prod file
  } else {
    // development
    require('dotenv').config({path: '.env'}); //to read info from .env file
  }
}

let path = require('path');

// --------------------------------------------------------
// --------------------------------------------------------
// See this issue here https://github.com/Ks89/My-MEAN-website/issues/30
//  to understand this piece of code.
let pathFrontEndFolder, pathFrontEndIndex;
let pathFrontEndAdminIndex;
if (config.isCI() || config.isTest()) {
  console.log(`Executed in CI or TEST - providing fake '../My-MEAN-website-client' and index.html`);
  //provides fake directories and files to be able to run this files
  //also with mocha in both testing and ci environments.
  //Otherwise, you are forced to run `npm run build` into ../My-MEAN-website-client's folder
  pathFrontEndFolder = path.join(__dirname);
  pathFrontEndIndex = path.join(__dirname, 'app.js');
} else {
  if (config.isProd()) {
    logger.warn('Providing both index.html and admin.html in a production environment');
    // you can add custom configuration here for production mode
  } else {
    logger.warn('Providing real ${config.FRONT_END_PATH}, index.html and admin.html');
  }
  pathFrontEndFolder = path.join(__dirname, config.FRONT_END_PATH);
  pathFrontEndIndex = path.join(__dirname, config.FRONT_END_PATH, 'index.html');
  pathFrontEndAdminIndex = path.join(__dirname, config.FRONT_END_PATH, 'admin.html');
}
// --------------------------------------------------------
// --------------------------------------------------------

let express = require('express');
// let vhost = require('vhost');
let compression = require('compression');
let morgan = require('morgan');
let session = require('express-session');
let bodyParser = require('body-parser');

let redis = require('redis'); //it's really useful?
let RedisStore = require('connect-redis')(session);
let client = redis.createClient(); //it's really useful?

// --------------------------------------------------------------------------
// ----------------------------security packages-----------------------------
// --------------------------------------------------------------------------
// All security features are prefixed with `--SEC--`
// --SEC-- - github analog-nico/hpp [NOT helmet]
//    [http params pollution] security package to prevent http params pollution
let hpp = require('hpp');
// --SEC-- - [CSRF] github.com/expressjs/csurf [NOT helmet]
let csrf = require('csurf');
// --SEC-- - authentication local/3dparty (OAuth)
let passport = require('passport');
// --SEC-- - github ericmdantas/express-content-length-validator [NOT helmet]
//    large payload attacks - Make sure this application is
//    not vulnerable to large payload attacks
let contentLength = require('express-content-length-validator');
const MAX_CONTENT_LENGTH_ACCEPTED = 9999; // constants used with `contentLength`
// --SEC-- - Helmet
let helmet = require('helmet');
// --------------------------------------------------------------------------
// --------------------------------------------------------------------------

logger.warn('Initializing mongodb');
//require for mongo
require('./src/models/db');
require('./src/controllers/authentication/passport')(passport);

logger.warn('Initializing expressjs');
let app = express();

logger.warn('Initializing helmet');
// --SEC-- - [helmet] enable helmet
// this automatically add 9 of 11 security features
/*
 -dnsPrefetchControl controls browser DNS prefetching
 -frameguard to prevent clickjacking
 -hidePoweredBy to remove the X-Powered-By header
 -hpkp for HTTP Public Key Pinning
 -hsts for HTTP Strict Transport Security
 -ieNoOpen sets X-Download-Options for IE8+
 -noSniff to keep clients from sniffing the MIME type
 -xssFilter adds some small XSS protections
 */
// The other features NOT included by default are:
/*
 -contentSecurityPolicy for setting Content Security Policy
 -noCache to disable client-side caching => I don't want this for better performances
 -referrerPolicy to hide the Referer header
 */
app.use(helmet());

// --SEC-- - hidePoweredBy: X-Powered-By forced to a fake value to
// hide the default 'express' value [helmet]
app.use(helmet.hidePoweredBy({setTo: config.HELMET_HIDE_POWERED_BY}));

// --SEC-- - noCache to disable client-side caching [helmet]
// I don't want this for better performances (leave commented :))
// app.use(helmet.noCache())

// --SEC-- - referrer-policy to hide the Referer header [helmet]
app.use(helmet.referrerPolicy({policy: config.HELMET_REFERRER_POLICY}));

// --SEC-- - Public Key Pinning (hpkp): HTTPS certificates can be forged,
//    allowing man-in-the middle attacks.
//    HTTP Public Key Pinning aims to help that. [helmet]
const ninetyDaysInSeconds = 7776000;
app.use(helmet.hpkp({
  maxAge: ninetyDaysInSeconds,
  sha256s: [config.HELMET_HPKP_SHA256S_1, config.HELMET_HPKP_SHA256S_2],
  includeSubdomains: true,         // optional
  reportUri: config.HELMET_HPKP_REPORT_URI, // optional
  reportOnly: false,               // optional
  // Set the header based on a condition.
  setIf: req => req.secure //optional ()
}));

// --SEC-- - Content Security Policy (CSP): Trying to prevent Injecting anything
//    unintended into our page. That could cause XSS vulnerabilities,
//    unintended tracking, malicious frames, and more. [helmet]
app.use(helmet.contentSecurityPolicy({
  directives: {
    defaultSrc: [`'self'`, 'localhost:3000', 'localhost:3001', 'www.google.com', 'www.youtube.com'],
    scriptSrc: [`'self'`, `'unsafe-inline'`, `'unsafe-eval'`, 'maxcdn.bootstrapcdn.com',
      'ajax.googleapis.com', 'cdnjs.cloudflare.com',
      'code.jquery.com', 'www.google.com',
      'www.gstatic.com'],
    styleSrc: [`'self'`, 'ajax.googleapis.com', 'maxcdn.bootstrapcdn.com', 'cdnjs.cloudflare.com', `'unsafe-inline'`],
    fontSrc: [`'self'`, 'maxcdn.bootstrapcdn.com'],
    imgSrc: [`'self'`, 'localhost:3000', 'localhost:3001',
      'placehold.it', 'placeholdit.imgix.net', 'camo.githubusercontent.com',
      's3.amazonaws.com', 'cdnjs.cloudflare.com'],
    sandbox: ['allow-forms', 'allow-scripts', 'allow-same-origin', 'allow-popups'],
    frameSrc: [`'self'`, 'www.google.com', 'www.youtube.com'], //frame-src is deprecated
    childSrc: [`'self'`, 'www.google.com', 'www.youtube.com'],
    connectSrc: [
      `'self'`, 'cdnjs.cloudflare.com', 'ajax.googleapis.com',
      'ws://localhost:3000', 'ws://localhost:3001', 'ws://localhost:3100',
      'ws://localhost:3300'
    ],
    reportUri: '/report-violation',
    objectSrc: [`'none'`]
  },
  // Set to true if you only want browsers to report errors, not block them
  reportOnly: false,
  // Set to true if you want to blindly set all headers: Content-Security-Policy,
  // X-WebKit-CSP, and X-Content-Security-Policy.
  setAllHeaders: false,
  // Set to true if you want to disable CSP on Android where it can be buggy.
  disableAndroid: false,
  // Set to false if you want to completely disable any user-agent sniffing.
  // This may make the headers less compatible but it will be much faster.
  // This defaults to 'true'.
  browserSniff: true
}));

// --SEC-- - large payload attacks:
//   this line enables the middleware for all routes [NOT helmet]
app.use(contentLength.validateMax({
  max: MAX_CONTENT_LENGTH_ACCEPTED,
  status: 400, message: config.LARGE_PAYLOAD_MESSAGE
})); // max size accepted for the content-length

logger.warn(`Initializing morgan (logger of req, res and so on... It's different from winston logger)`);
if (!config.isCI() && !config.isTest()) {
  // Disable morgan while testing to prevent very big log with useless information
  app.use(morgan('combined', {'stream': logger.stream}));
}

logger.warn('Initializing static resources');
app.use(express.static(pathFrontEndFolder));

logger.warn('Initializing bodyparser');

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({extended: false}));

logger.warn('Initializing hpp');
// --SEC-- - http params pollution: activate http parameters pollution
// use this ALWAYS AFTER app.use(bodyParser.urlencoded()) [NOT helmet]
app.use(hpp());

logger.warn('Initializing Express session');
// Express Session
app.use(session({
  secret: config.EXPRESS_SESSION_SECRET,
  resave: false,
  saveUninitialized: true,
  store: new RedisStore({host: config.REDIS_HOST, port: config.REDIS_PORT, client: client, ttl: config.REDIS_TTL})
  // cookie: {
  //   httpOnly: false,
  //     secure: false, //to use true, you must use https. If you'll use true with http it won't work.
  //     //maxAge: 2160000000
  // }
}));

logger.warn('Initializing passportjs');

app.use(passport.initialize());
app.use(passport.session());

// compress all requests using gzip
app.use(compression());

logger.warn('Initializing REST apis and CSRF');

// --------------------------------------- ROUTES ---------------------------------------
// dedicated routes for angular logging with stacktracejs
// these router aren't protected with csrf, because declared before app.use(csrf()).
let loggerApi = require('./src/routes/log-api')(express);
app.use('/api/log', loggerApi);

// enable middleware CSRF by csurf package [NOT helmet]
// before app.use('/api', routesApi); to protect their,
// but after session and/or cookie initialization
app.use(csrf());
app.use(function (req, res, next) {
  res.cookie('XSRF-TOKEN', req.csrfToken());
  res.locals.csrftoken = req.csrfToken();
  next();
});

// APIs for all route protected with CSRF (all routes except for angular log's service)
let routesApi = require('./src/routes/index')(express);
app.use('/api', routesApi);
// --------------------------------------------------------------------------------------

logger.warn('Initializing static path for both index.html and admin.html');

app.use('/', function (req, res) {
  res.sendFile(pathFrontEndIndex);
});

app.use('/admin', function (req, res) {
  res.sendFile(pathFrontEndAdminIndex);
});

// app.use(vhost('mymeanwebsite.it', function (req, res) {
//   res.sendFile(pathFrontEndIndex);
// }));
//
// app.use(vhost('admin.mymeanwebsite.it', function (req, res) {
//   res.sendFile(pathFrontEndAdminIndex);
// }));


// catch bad csrf token
app.use(function (err, req, res, next) {
  if (err.code !== 'EBADCSRFTOKEN') {
    return next(err);
  }
  // handle CSRF token errors here
  res.status(403);
  res.json({ message: 'session has expired or form tampered with'});
});


// catch 404 and forward to error handler
app.use(function (req, res, next) {
  let err = new Error('Not Found');
  err.status = 404;
  next(err);
});

// error handlers
// Catch unauthorised errors
app.use(function (err, req, res) {
  if (err.name === 'UnauthorizedError') {
    res.status(401);
    res.json({ message: `${err.name}: ${err.message}`});
  }
});

// development error handler
// will print stacktrace
if (app.get('env') === 'development') {
  app.use(function (err, req, res) {
    res.status(err.status || 500);
    res.render('error', {
      message: err.message,
      error: err
    });
  });
}

// production error handler
// no stacktraces leaked to user
app.use(function (err, req, res) {
  res.status(err.status || 500);
  res.render('error', {
    message: err.message,
    error: {}
  });
});

module.exports = app;